src/SqlSquare/AST.purs

```diff
@@ -1,225 +1,125 @@
-  IdentRelation { ident, alias } →
-    ident <> F.foldMap (" as " <> _) alias
-  JoinRelation { left, right, joinType, clause } →
-    printRelation left
-    <> " "
-    <> printJoinType joinType
-    <> " "
-    <> printRelation right
-    <> " on "
-    <> clause
+import Data.Ord (class Ord1)
+import Data.String.Regex as RX
+import Data.String.Regex.Flags as RXF
+import Data.String.Regex.Unsafe as URX
+
+import SqlSquare.Utils (type (×), (×), (∘), (⋙))
+import SqlSquare.OrderType as OT
+import SqlSquare.JoinType as JT
+import SqlSquare.BinaryOperator (BinaryOperator(..))
+import SqlSquare.UnaryOperator (UnaryOperator(..))
+import SqlSquare.GroupBy (GroupBy(..), printGroupBy)
+import SqlSquare.Case (Case(..), printCase)
+import SqlSquare.OrderBy (OrderBy(..), printOrderBy)
+import SqlSquare.Projection (Projection(..), printProjection)
+import SqlSquare.Relation (Relation(..), printRelation, FUPath, JoinRelR, ExprRelR, TableRelR, VariRelR, IdentRelR)
+
+import Matryoshka (class Recursive, Algebra, cata, transParaT, project)
+
+import Debug.Trace as DT
+
+type BinopR a =
+  { lhs ∷ a
+  , rhs ∷ a
+  , op ∷ BinaryOperator
+  }
+
+type UnopR a =
+  { expr ∷ a
+  , op ∷ UnaryOperator
+  }
+
+type InvokeFunctionR a =
+  { name ∷ String
+  , args ∷ L.List a
+  }
+
+type MatchR a =
+  { expr ∷ a
+  , cases ∷ L.List (Case a)
+  , else_ ∷ Maybe a
+  }
+
+type SwitchR a =
+  { cases ∷ L.List (Case a)
+  , else_ ∷ Maybe a
+  }
+
+type LetR a =
+  { ident ∷ String
+  , bindTo ∷ a
+  , in_ ∷ a
+  }
+
+type SelectR a =
+  { isDistinct ∷  Boolean
+  , projections ∷ L.List (Projection a)
+  , relations ∷ Maybe (Relation a)
+  , filter ∷ Maybe a
+  , groupBy ∷ Maybe (GroupBy a)
+  , orderBy ∷ Maybe (OrderBy a)
+  }
 
 data SqlF a
-  = SetLiteral (List a)
-  | ArrayLiteral (List a)
-  | MapLiteral (List (a × a))
+  = SetLiteral (L.List a)
+  | ArrayLiteral (L.List a)
+  | MapLiteral (L.List (a × a))
   | Splice (Maybe a)
-  | Binop
-      { lhs ∷ a
-      , rhs ∷ a
-      , op ∷ BinaryOperator
-      }
-  | Unop
-      { expr ∷ a
-      , op ∷ UnaryOperator
-      }
+  | Binop (BinopR a)
+  | Unop (UnopR a)
   | Ident String
-  | InvokeFunction
-      { name ∷ String
-      , args ∷ List a
-      }
-  | Match
-      { expr ∷ a
-      , cases ∷ List (Case a)
-      , default_ ∷ Maybe a
-      }
-  | Switch
-      { cases ∷ List (Case a)
-      , default_ ∷ Maybe a
-      }
-  | Let
-      { ident ∷ String
-      , bindTo ∷ a
-      , in_ ∷ a
-      }
+  | InvokeFunction (InvokeFunctionR a)
+  | Match (MatchR a)
+  | Switch (SwitchR a)
+  | Let (LetR a)
   | IntLiteral Int
   | FloatLiteral Number
   | StringLiteral String
   | NullLiteral
   | BoolLiteral Boolean
   | Vari String
-  | Select
-      { isDistinct ∷ Boolean
-      , projections ∷ List (Projection a)
-      , relations ∷ Maybe (SqlRelation a)
-      , filter ∷ Maybe a
-      , groupBy ∷ Maybe (GroupBy a)
-      , orderBy ∷ Maybe (OrderBy a)
-      }
+  | Select (SelectR a)
+  | Parens a

```

Comment:
Yes, using EJson is a good idea. Technically you could use `Coproduct SqlF EJsonF` where `SqlF` is the part of SQL<sup>2</sup> that is not `EJson`.

Remember in SQL<sup>2</sup> you can do things like this:

```
SELECT {d.dynamic_field : [d.another_one]} FROM `/foo/` AS d
```

i.e. "literals" can recursively contain arbitrary SQL<sup>2</sup> / EJSON.